File: subprojects/language-web/src/main/java/tools/refinery/language/web/api/dto/Scope.java
/*
 * SPDX-FileCopyrightText: 2025 The Refinery Authors <https://refinery.tools/>
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package tools.refinery.language.web.api.dto;

public class Scope {
	private String relation;
	private boolean override;
	private boolean incremental;
	private int lowerBound;
	private Integer upperBound;

	public String getRelation() {
		return relation;
	}

	public void setRelation(String relation) {
		this.relation = relation;
	}

	public boolean isIncremental() {
		return incremental;
	}

	public boolean isOverride() {
		return override;
	}

	public void setOverride(boolean override) {
		this.override = override;
	}

	public void setIncremental(boolean incremental) {
		this.incremental = incremental;
	}

	public int getLowerBound() {
		return Math.max(0, lowerBound);
	}

	public void setLowerBound(int lowerBound) {
		this.lowerBound = lowerBound;
	}

	public Integer getUpperBound() {
		return upperBound == null ? null : Math.max(0, upperBound);
	}

	public void setUpperBound(Integer upperBound) {
		this.upperBound = upperBound;
	}

	public String toScopeConstraint() {
		return relation +
				(incremental ? "+=" : "=") +
				lowerBound +
				".." +
				(upperBound == null ? "*" : upperBound);
	}
}

File: subprojects/client-js/src/dto/Scope.ts
/*
 * SPDX-FileCopyrightText: 2025 The Refinery Authors <https://refinery.tools/>
 *
 * SPDX-License-Identifier: EPL-2.0
 */

import z from 'zod/v4';

export const Scope = z
  .object({
    relation: z.string(),
    override: z.boolean().default(false),
    incremental: z.boolean().default(false),
    lowerBound: z.number().nonnegative().default(0),
    upperBound: z.number().nonnegative().optional(),
  })
  .refine(
    ({ lowerBound, upperBound }) =>
      upperBound === undefined || lowerBound <= upperBound,
    {
      message: 'lowerBound must be less than or equal to upperBound',
    },
  );

export type Scope = z.infer<typeof Scope>;
